### PR TITLE
`__osMalloc.c` OK

### PR DESCRIPTION
**Pull request by AngheloAlf**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

Joining to the `boot` wagon 

I noticed this file wasn't completely decomped (2 functions were missing, bss and unreferenced strings wasn't imported), so I decided to look at it.
The functions weren't hard. I decided to cleanup the file a bit too

### CODE REVIEW

File: src/boot_O2/__osMalloc.c
Comment (EllipticEllipsis) [inline src/boot_O2/__osMalloc.c:63]: Sounds like we need `ptrdiff_t` as well.

File: src/boot_O2/__osMalloc.c
Comment (EllipticEllipsis) [inline src/boot_O2/__osMalloc.c:64]: Can this have a more descriptive name? What's its purpose?

File: src/boot_O2/__osMalloc.c
Comment (EllipticEllipsis) [inline src/boot_O2/__osMalloc.c:230]: Interesting, I thought realloc usually required a prior malloc.

File: src/boot_O2/__osMalloc.c
Comment (EllipticEllipsis) [inline src/boot_O2/__osMalloc.c:246]: Since this is the fancy bit of realloc, could you annotate what this is doing a bit?

File: src/boot_O2/__osMalloc.c
Comment (EllipticEllipsis) [inline src/boot_O2/__osMalloc.c:313]: Now where have I seen this style of error message before? 🤔

**Review by EllipticEllipsis — COMMENTED** [review]
Code looks good, but I think this file could do with some more general documentation given its importance. I've pointed out a few places, but there's more I'm sure.

**Comment by AngheloAlf** [comment]
This will be a draft until #396 is merged

File: src/boot_O2/__osMalloc.c
Comment (Kenix3) [inline src/boot_O2/__osMalloc.c:55]:
```suggestion
void __osMallocAddBlock(Arena* arena, void* start, size_t size) {
```

**Review by Kenix3 — CHANGES_REQUESTED** [review]

File: src/boot_O2/__osMalloc.c
Comment (AngheloAlf) [reply src/boot_O2/__osMalloc.c:230]: Me too, but it looks like this is standard behaviour.

> If ptr is NULL, the behavior is the same as calling malloc(new_size). 

https://en.cppreference.com/w/c/memory/realloc

**Comment by AngheloAlf** [comment]
Since the other PR was merged, I undrafted this PR and addressed the reviews.
I also added a `os_malloc.h` header to move those specific struct definitions out of `z64.h`

**Review by Kenix3 — APPROVED** [review]

**Review by hensldm — APPROVED** [review]
LGTM

**Comment by Kenix3** [comment]
Merge pending @EllipticEllipsis or @Rozelette approval.

File: src/boot_O2/__osMalloc.c
Comment (EllipticEllipsis) [inline src/boot_O2/__osMalloc.c:318]: This looks like an unordered list.

File: src/boot_O2/__osMalloc.c
Comment (EllipticEllipsis) [inline src/boot_O2/__osMalloc.c:322]: Might this also be clearer as an unordered list?

File: src/boot_O2/__osMalloc.c
Comment (EllipticEllipsis) [inline src/boot_O2/__osMalloc.c:328]:
```suggestion
 * and the original parameter \p ptr remains valid.
```

File: src/boot_O2/__osMalloc.c
Comment (EllipticEllipsis) [inline src/boot_O2/__osMalloc.c:390]:
```suggestion
 * @brief Gets the size of the biggest block which isn't allocated, the total free space and the total allocated space.
```

File: src/boot_O2/__osMalloc.c
Comment (EllipticEllipsis) [inline src/boot_O2/__osMalloc.c:393]:
```suggestion
 * @param[out] outMaxFree  The value of the biggest block which is not allocated.
```

File: src/boot_O2/__osMalloc.c
Comment (EllipticEllipsis) [inline src/boot_O2/__osMalloc.c:395]: Maybe "total" instead of "accumulated"? Is there a subtlety I'm missing?

**Review by EllipticEllipsis — COMMENTED** [review]
Wow, that's a lot of good documentation. I especially like the Doxygen bits, although I wonder if this is how to use it to its maximum potential. Just a few minor grammar/formatting things now.

File: src/boot_O2/__osMalloc.c
Comment (AngheloAlf) [reply src/boot_O2/__osMalloc.c:318]: I blame the formater for messing up this

File: src/boot_O2/__osMalloc.c
Comment (EllipticEllipsis) [inline src/boot_O2/__osMalloc.c:133]: Basically this is the same thing, right?

Comment (AngheloAlf) [reply src/boot_O2/__osMalloc.c:133]: Yes...

**Review by EllipticEllipsis — APPROVED** [review]

File: src/boot_O2/__osMalloc.c
Comment (Thar0) [inline src/boot_O2/__osMalloc.c:57]:
```suggestion
 * @brief Initializes \p arena to manage the memory region \p heap.
 *
 * @param arena  The Arena to initialize.
 * @param heap   The memory region to use as heap space.
 * @param size   The size of the heap.
```

File: src/boot_O2/__osMalloc.c
Comment (Thar0) [inline src/boot_O2/__osMalloc.c:68]: This function initializes the first block, "add block" suggests to me that it's used whenever a new block is added, which is not the case. I'm not totally sure on a succinct name however.

File: src/boot_O2/__osMalloc.c
Comment (Thar0) [inline src/boot_O2/__osMalloc.c:127]:
```suggestion
 * The block of memory will be allocated in the start of the first sufficiently large free block.
```
Maybe this is implied? I feel like it helps to be specific here

File: src/boot_O2/__osMalloc.c
Comment (Thar0) [inline src/boot_O2/__osMalloc.c:192]:
```suggestion
 * Unlike __osMalloc, the block of memory will be allocated from the end of the \p arena.
```
In a similar vein to above, "at" suggests to me at first glance that it would be placed at the very end of the arena, rather than the last sufficiently large free block.

File: src/boot_O2/__osMalloc.c
Comment (Thar0) [inline src/boot_O2/__osMalloc.c:260]: `the pointer \p ptr` feels unnecessarily verbose, does it really need to be stated explicitly that `ptr` is a pointer with each mention?

File: src/boot_O2/__osMalloc.c
Comment (Thar0) [inline src/boot_O2/__osMalloc.c:391]:
```suggestion
 * @brief Gets the size of the biggest block which isn't allocated, the total free space and the total allocated space.
```
```suggestion
 * @brief Gets the size of the largest free block, the total free space and the total allocated space.
```

File: src/boot_O2/__osMalloc.c
Comment (Thar0) [inline src/boot_O2/__osMalloc.c:394]:
```suggestion
 * @param[out] outMaxFree  The size of the largest free block.
```

**Review by Thar0 — COMMENTED** [review]
Some doc comments

File: src/boot_O2/__osMalloc.c
Comment (AngheloAlf) [reply src/boot_O2/__osMalloc.c:68]: You are right.
But at the same time, this function _has_ the functionality to add a new block, so the name isn't technically wrong, it just happens that each an allocating function "adds" a block, it does it manually by spliting one of the already existing ones.

This function could theoretically be used to add blocks new blocks to completely unrelated heaps, it just happens that the game never does this.

What about `__osMallocAddHeap`?

Comment (AngheloAlf) [reply src/boot_O2/__osMalloc.c:68]: Also, this name comes from the Animal Forest map.

I took a quick look at AF and I was able to find this function (ROM 0x008060, VRAM: 0x8002CC60 for anybody interested). It is exactly the same (and it matches without -g3. https://decomp.me/scratch/z9aiE).

There's a catch tho. I compared the other functions in AF, and looks like there are a few functions in the AF map missing from the AF rom, which shouldn't happen, right? Also, the VRAMs from the map doesn't match the VRAMs from the ROM itself (which may not be a huge deal? idk).

My conclusion is, what we have been calling `__osMalloc` all this time is not the real `__osMalloc` that the function map refers to. If this is true, I propose to rename all the functions in this file to something more ad hoc to our naming scheme (`SystemName_Func`). What do you think?

Comment (Thar0) [reply src/boot_O2/__osMalloc.c:68]: Perhaps I'm wrong, but I'm pretty sure this cannot add a new block, only the first block, which is an important distinction. It looks to me that if this function were called more than once for a particular Arena, the arena would lose the ability to keep track of prior blocks in the other heap space. If that's correct, I suggest `__osMallocAddFirstBlock`. Otherwise, I think `__osMallocAddHeap` is good.

Regarding the AF map, the map comes from the port to Gamecube, so there are bound to be a number of differences when looking at the N64 version. It's likely AF on the Gamecube uses a modified `__osMalloc` but it should certainly have it's roots in the N64 version.
I see that `__osMallocAddBlock` is the original name as you pointed out, which is surely why it was named that way in the first place. I don't believe we should stick to it though if it's not representative of the purpose of the function, which I don't think it really is. Perhaps it would be nice however to comment it like
`// Original name: __osMallocAddBlock`

Comment (Thar0) [reply src/boot_O2/__osMalloc.c:68]: Ah I see what you mean now, I failed to notice that it does in fact link with existing blocks via `ArenaImpl_GetLastBlock`. In that case, I think `__osMallocAddHeap` is suitable.

Comment (AngheloAlf) [reply src/boot_O2/__osMalloc.c:68]: Sounds good. I'll rename it to `__osMallocAddHeap`

File: src/boot_O2/__osMalloc.c
Comment (hensldm) [inline src/boot_O2/__osMalloc.c:68]:
```suggestion
// Original name: __osMallocAddBlock
```

**Review by Thar0 — APPROVED** [review]

**Pull request merged**
